Deriving Deserialize for structs with Cow<'a, str> fails

Consider the following code:
```rust
#[derive(Deserialize)]
struct S<'a>(Cow<'a, str>);
```
This fails with a cryptic error message:
```
error[E0207]: the lifetime parameter `'a` is not constrained by the impl trait, self type, or predicates
 --> <quote expansion>:1:2
  |
1 | <'a>
  |  ^^ unconstrained lifetime parameter
```
I believe `Deserialize` is implemented for `Cow<'a, str>`, so I think this should work.

--- a/serde_codegen/src/de.rs
+++ b/serde_codegen/src/de.rs
@@ -192,7 +192,7 @@ fn deserialize_visitor(
     builder: &aster::AstBuilder,
     generics: &ast::Generics,
 ) -> (P<ast::Item>, P<ast::Ty>, P<ast::Expr>) {
-    if generics.ty_params.is_empty() {
+    if generics.lifetimes.is_empty() && generics.ty_params.is_empty() {
         (
             builder.item().unit_struct("__Visitor"),
             builder.ty().id("__Visitor"),

--- a/testing/tests/test_gen.rs
+++ b/testing/tests/test_gen.rs
@@ -6,6 +6,7 @@ extern crate serde;
 use self::serde::ser::{Serialize, Serializer};
 use self::serde::de::{Deserialize, Deserializer};
 
+use std::borrow::Cow;
 use std::marker::PhantomData;
 
 //////////////////////////////////////////////////////////////////////////
@@ -177,6 +178,15 @@ fn test_gen() {
         e: E,
     }
     assert::<WithTraits2<X, X>>();
+
+    #[derive(Serialize, Deserialize)]
+    struct CowStr<'a>(Cow<'a, str>);
+    assert::<CowStr>();
+
+    #[derive(Serialize, Deserialize)]
+    #[serde(bound(deserialize = "T::Owned: Deserialize"))]
+    struct CowT<'a, T: ?Sized + 'a + ToOwned>(Cow<'a, T>);
+    assert::<CowT<str>>();
 }
 
 //////////////////////////////////////////////////////////////////////////

